source/grid/grid_tools.cc

```diff
@@ -3035,6 +3036,76 @@ namespace GridTools
   }
 
 
+
+  template <int dim, int spacedim>
+  std::vector<types::subdomain_id>
+  get_subdomain_association(const Triangulation<dim, spacedim> &triangulation,
+                            const std::vector<CellId> &         cell_ids)
+  {
+    std::vector<types::subdomain_id> subdomain_ids;
+    subdomain_ids.reserve(cell_ids.size());
+
+    if (dynamic_cast<
+          const parallel::fullydistributed::Triangulation<dim, spacedim> *>(
+          &triangulation) != nullptr)
+      {
+        Assert(false, ExcNotImplemented());
+      }
+    else if (const parallel::distributed::Triangulation<dim, spacedim>
+               *parallel_tria =
+                 dynamic_cast<
+                   const parallel::distributed::Triangulation<dim, spacedim> *>(
+                   &triangulation) != nullptr)
+      {
+        // for parallel distributed triangulations, we will ask the p4est oracle
+        // about the global partitioning of active cells since this information
+        // is stored on every processor
+        for (const auto &cell_id : cell_ids)
+          {
+            // find descendent from coarse quadrant
+            typename dealii::internal::p4est::types<dim>::quadrant p4est_cell,
+              p4est_children[GeometryInfo<dim>::max_children_per_cell];
+
+            dealii::internal::p4est::init_coarse_quadrant<dim>(p4est_cell);
+            for (const auto &child_index : cell_id.get_child_indices())
+              {
+                dealii::internal::p4est::init_quadrant_children<dim>(
+                  p4est_cell, p4est_children);
+                p4est_cell =
+                  p4est_children[static_cast<unsigned int>(child_index)];
+              }
+
+            // find owning process, i.e., the subdomain id
+            const int owner =
+              dealii::internal::p4est::functions<dim>::comm_find_owner(
+                *(parallel_tria->get_p4est()),
+                cell_id.get_coarse_cell_id(),
+                &p4est_cell,
+                Utilities::MPI::this_mpi_process(
+                  parallel_tria->get_communicator()));
+
+            Assert(owner >= 0, ExcMessage("p4est should know the owner."));
+
+            subdomain_ids.push_back(owner);
+          }
+      }
+    else
+      {
+        // for serial and parallel shared triangulations, information about the
+        // subdomain association of all active cells is available on every
+        // processor
+        for (const auto &cell_id : cell_ids)
+          {
+            subdomain_ids.push_back(
+              triangulation.create_cell_iterator(cell_id).get_subdomain_id());
+          }
+      }
```

Comment:
This won't work for a `p:s:T`s. In the case that artificial cells are supported, you need to load the subdomain ids like this: 

https://github.com/dealii/dealii/blob/b5f7124da5782f93de6add00f177f7869bbe9cef/source/distributed/shared_tria.cc#L188-L196

and clear them again:

https://github.com/dealii/dealii/blob/b5f7124da5782f93de6add00f177f7869bbe9cef/source/distributed/shared_tria.cc#L262-L270

Reply:
Ah yes, you are right. We need to use `true_subdomain_ids` for this. I will make a `p::s::T` specialization after lunch :)